mariadb: backport fixes for Linux in latest version

danielnachun opened the pull request:
- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/linuxbrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/linuxbrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?
- [ ] Have you included the output of `brew gist-logs <formula>` of the build failure if your PR fixes a build failure. Please quote the exact error message.

-----

This is a backport of the fixes I made in https://github.com/Homebrew/linuxbrew-core/pull/21644 to all other versions of `mariadb`.  Along the way, I backported the other changes that were in the latest version of `mariadb` but not the older ones.

gromgit commented on Formula/mariadb@10.1.rb:
```diff
@@ -111,7 +130,7 @@ def install
 
   def post_install
-    return if ENV["CI"]
+    return if ENV["CI"] && !OS.mac?
```
This shouldn't be necessary, as the macOS formulae now take the same action.

gromgit commented on Formula/mariadb@10.1.rb:
```diff
@@ -161,6 +180,8 @@ def plist
   end
 
   test do
+    return if ENV["CI"] && !OS.mac?
```
Suggest looking at https://github.com/Homebrew/homebrew-core/pull/66727 and transplanting the new test case I created for all the `mariadb` formulae.

danielnachun commented:
@gromgit Thanks for pointing me to that PR, I merged that into my branch but clearly got confused about what was new and what was supposed to remain the same.  I will fix everything to match what you've already done.  

The main purpose of this PR was to disable `JDBC` and `DTRACE`, and in the process of doing so I noticed some other dependency-related Linux fixes needed to be backported as well.  I'm glad to see that the `post_install` and `test` blocks are already good to go!

danielnachun commented:
I think I got all the `test` and `post_install` blocks back to the way they should be, but now the tests fail when calling `mysqld`.  If I understand the [knowledge base](https://mariadb.com/kb/en/running-mysqld-as-root/) right, `mysqld` needs `user=root` when run as root, and `ENV["USER"]` is `root` in CI.  Should I add `user=ENV["USER"]` as an argument to `mysqld` in that line of the test?  @gromgit let me know what you think the best option is.

gromgit commented:
Try your method first, and see if it works.

`ENV["USER"]==root` on Linux CI is just...strange. Can you add the following to the start of the test block, to see if it's really running as root?
```
ohai shell_output("id")
```

gromgit commented:
Actually, try adding:
```
ENV["USER"] = shell_output("id -un").chomp
```
at the start of the test block.

danielnachun commented:
I added `ENV["USER"] = shell_output("id -un").chomp`, but I guess `id -un` must return `root` because that was the value returned by `ENV["USER"]` when calling `install_db`.

On the upside, adding `user=ENV["USER"]` to the call to `mysqld` seems to make the tests pass!  If we think this is the way to go forward, should I open PR's to fix this for all versions of `mysql` and `percona`?  I'm also assuming this should cause no issues for macOS but let me know if I should test that as well.

gromgit commented:
Actually, in hindsight, not starting `mysqld` with `--user` is the real problem, so you should be able to drop the `id -un` line. Probably a good idea to standardize on that across all the MySQL-ish formulae, so please do the same for the other formulae on this side, and I'll take care of the macOS side.

danielnachun commented:
I removed the `id -un` line from the formulae, and I also realized that the current `mariadb` needs to be tweaked as well, so I added that to this PR as well.  I will open a separate PR for `mysql` and `persona` to avoid confusion here.

danielnachun commented:
Of course adding the current `mariadb` formula uncovered a new problem.  It looks like it is linking to the wrong `libstdc++`, which I've run into before.  I don't know why it's only affecting `mariadb` and not the other versions.

danielnachun commented:
`mecab` needed to be built with `gcc@7` for the current `mariadb` formula to work.  Everything passes CI so I think this is ready to go!

danielnachun commented:
This PR has been a bit of a roller coaster.  I've rebased to integrate the test fixes @gromgit added to homebrew-core.  Now the only changes are to build with `gcc@7`, add some missing dependencies and disable some features so that all versions will build on Linux.  The current `mariadb` formula needed some slight tweaks, and I also had to make `mecab` build with `gcc@7` for `mariadb` to pass testing.

jonchang commented:
Tests for `mecab` and `mariadb` are failing due to existing bottles for Linux, but these should be kept. Gonna "rebase and merge" this then `brew request-bottle mariadb@10.1 mariadb@10.2 mariadb@10.3 mariadb@10.4`.

danielnachun commented:
I changed everything over to `on_linux` blocks and it looks good now.

jonchang commented:
Great work, thanks for sticking with this @danielnachun!

danielnachun commented:
It looks like the bottling of the legacy versions failed.  In examining the errors, I see they have a couple of different causes.  In `mariadb@10.1`, I see that the shim cleanup needs to be run on Linux as well.  I just had to fix this in another formula, so I will open a new PR to fix that.

For `mariadb@10.4`, there is an issue with the upstream URL for the source tarball.  I tried to access it in my web browser, and it just hangs.  I'm hoping this is just a transient issue.  

I don't know if this will affect `mariadb@10.4` because it didn't download, but the other 3 versions also complained about a non-executable in the `bin` folder.  The offending file is in `$HOMEBREW_PREFIX/opt/mariadb@VERSION/bin/rcmysql`, and it is indeed just a symlink to a file in `$HOMEBREW_PREFIX/opt/mariadb@VERSION/etc`.  Is that an error we need to worry about?  I'm assuming that was put there for a reason (maybe backwards compatibility with MySQL?), but I don't know enough about MariaDB/MySQL to be sure.
